File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return head;
        }

        ListNode prevNode = null;
        ListNode currentNode = head;
        ListNode tempNode = null;
        
        while (true) {
            tempNode = currentNode.next;
            currentNode.next = prevNode;
            prevNode = currentNode;
            if (tempNode == null) {
                break;
            }
            currentNode = tempNode;
        }
        return currentNode;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head, n) {
        let reversedHead = this.reverseList(head);
        let copy = reversedHead;

        if (n == 1) {
            return reversedHead.next ? this.reverseList(reversedHead.next) : null;
        }

        while (true) {
            if ((--n - 1) > 0) {
                copy = copy.next;
            } else {
                break;
            }
        }

        if (copy?.next) {
            copy.next = copy.next.next;
        }

        return this.reverseList(reversedHead);
    }

    reverseList(node) {
        let prevNode = null;
        let currentNode = node;
        let tempNode = null;
        
        while (true) {
            tempNode = currentNode.next;
            currentNode.next = prevNode;
            prevNode = currentNode;
            if (tempNode == null) {
                break;
            }
            currentNode = tempNode;
        }

        return currentNode;
    }
}
